cobra/core/metabolite.py

```diff
@@ -131,39 +131,29 @@ def y(self):
         the solution.
 
         """
-        warn("use metabolite.shadow_price instead", DeprecationWarning)
-        try:
-            return self._model.solution.y_dict[self.id]
-        except Exception as e:
-            if self._model is None:
-                raise Exception("not part of a model")
-            if not hasattr(self._model, "solution") or \
-                    self._model.solution is None or \
-                    self._model.solution.status == "NA":
-                raise Exception("model has not been solved")
-            if self._model.solution.status != "optimal":
-                raise Exception("model solution was not optimal")
-            raise e  # Not sure what the exact problem was
+        warn("Please use metabolite.shadow_price instead.", DeprecationWarning)
+        return self.shadow_price
 
     def shadow_price(self):
-        """The shadow price for the metabolite in the most recent solution
+        """The shadow price for the metabolite in the most recent solution.
 
-        Shadow prices are computed from the dual values of the bounds in
-        the solution.
+        Shadow prices are computed from the dual values of the model constraints
+        in the solution.
 
+        Warning
+        -------
+        You are responsible for checking the solver status. You might be
+        interested in values of non-optimal solutions
         """
         try:
-            return self._model.solution.shadow_prices[self.id]
-        except Exception as e:
-            if self._model is None:
-                raise Exception("not part of a model")
-            if not hasattr(self._model, "solution") or \
-                    self._model.solution is None or \
-                    self._model.solution.status == "NA":
-                raise Exception("model has not been solved")
-            if self._model.solution.status != "optimal":
-                raise Exception("model solution was not optimal")
-            raise e
+            if self._model.solver.status != "optimal":
+                warn("Solver status is not optimal ({}), please treat value"
```

Comment:
I haven't tested but I was always under the impression that repeated exact same warnings are swallowed whereas log output is always written (unless filtered by level). I could also remove the check completely and make it a further comment in the docstring.